pkg/tanka/export.go

```diff
@@ -0,0 +1,209 @@
+package tanka
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+
+	"github.com/grafana/tanka/pkg/kubernetes/manifest"
+	"github.com/grafana/tanka/pkg/process"
+)
+
+// BelRune is a string of the Ascii character BEL which made computers ring in ancient times
+// We use it as "magic" char for the subfolder creation as it is a non printable character and thereby will never be
+// in a valid filepath by accident. Only when we include it.
+const BelRune = string(rune(7))
+
+// When exporting manifests to files, it becomes increasingly hard to map manifests back to its environment, this file
+// can be used to map the files back to their environment. This is aimed to be used by CI/CD but can also be used for
+// debugging purposes.
+const manifestFile = "manifest.json"
+
+type ExportEnvOpts struct {
+	Format    *string   // formatting the filename based on the exported Kubernetes manifest
+	Extension *string   // extension of the filename
+	Merge     *bool     // merge export with existing directory
+	Targets   []string  // optional: only export specified Kubernetes manifests
+	ParseOpts ParseOpts // optional: options for parsing Environments
+}
+
+func DefaultExportEnvOpts() ExportEnvOpts {
+	format := "{{env.spec.namespace}}/{{env.metadata.name}}/{{.apiVersion}}.{{.kind}}-{{.metadata.name}}"
+	extension := "yaml"
+	merge := false
+	return ExportEnvOpts{
+		Format:    &format,
+		Extension: &extension,
+		Merge:     &merge,
+	}
+}
+
+func ExportEnvironments(paths []string, to string, opts *ExportEnvOpts) error {
+	// Keep track of which file maps to which environment
+	fileToEnv := map[string]string{}
+
+	// dir must be empty
+	empty, err := dirEmpty(to)
+	if err != nil {
+		return fmt.Errorf("Checking target dir: %s", err)
+	}
+	if !empty && !*opts.Merge {
+		return fmt.Errorf("Output dir `%s` not empty. Pass --merge to ignore this", to)
+	}
+
+	// get all environments for paths
+	envs, err := ParseEnvs(paths, opts.ParseOpts)
+	if err != nil {
+		return err
+	}
+
+	for _, env := range envs {
+		// select targets to export
+		filter, err := StringsToRegexps(opts.Targets)
+		if err != nil {
+			return err
+		}
+
+		// get the manifests
+		res, err := LoadManifests(env, filter)
+		if err != nil {
+			return err
+		}
+
+		// create raw manifest version of env for templating
+		raw, err := json.Marshal(env)
+		if err != nil {
+			return err
+		}
+		var menv manifest.Manifest
+		if err := json.Unmarshal(raw, &menv); err != nil {
+			return err
+		}
```

Comment:
This does heavy re-serialization of all k8s objects in the env, which quickly becomes inefficient. Let's exclude this data